Save attached outputs for compile-only cache entries

cowwoc opened the pull request:
## Summary
- reset cached-output bookkeeping per save invocation and always attach configured directories before writing the cache entry
- ensure compile-only executions persist the configured outputs so higher lifecycle phases can restore them
- add an integration test project covering JPMS module descriptors after a compile-only cache hit

## Testing
- ./mvnw -Prun-its -Dit.test=Issue393CompileRestoreTest verify

Fixes #393
Relates-To: #259
Relates-To: #340

elharo reviewed (changes requested):
need to run mvn spotless:apply

olamy commented on pom.xml:
```diff
@@ -427,6 +428,17 @@ under the License.
+        <configuration>
+          <java>
+            <palantirJavaFormat>
+              <version>2.81.0</version>
```
why do we need this here? It should come from parent pom

cowwoc replied:
Agreed, but how/where do I update the parent pom?

olamy replied:
what is the reason of such change?

Bukama replied:
In maven-parent `<version.palantirJavaFormat>2.80.0</version.palantirJavaFormat>` is defined so we have a consistent version across all projects.

cowwoc replied:
@olamy The version that is used by `maven-parent` is incompatible with JDK 25 which I was using locally. That said, I'll remove the change from the PR.
@Bukama Thanks

cowwoc commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ out/
 .java-version
 .checkstyle
 .factorypath
+null/
```
Oops :) Removed

erik-meuwese-topicus commented on src/test/projects/issue-393-compile-restore/consumer/src/test/java/org/apache/maven/caching/test/jpms/consumer/ConsumerTest.java:
```diff
@@ -0,0 +1,13 @@
+package org.apache.maven.caching.test.jpms.consumer;
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.maven.caching.test.jpms.consumer;
```

The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

erik-meuwese-topicus commented on src/test/projects/issue-393-compile-restore/app/src/main/java/module-info.java:
```diff
@@ -0,0 +1,3 @@
+module org.apache.maven.caching.test.jpms.app {
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
module org.apache.maven.caching.test.jpms.app {
```

The code convention requires the ASF license header
https://maven.apache.org/developers/conventions/code.html

cowwoc commented on src/test/projects/git-checkout-stale-artifact/pom.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Fixed, thank you!

cowwoc commented on src/test/projects/git-checkout-stale-multimodule/.mvn/maven-build-cache-config.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Fixed, thank you!

cowwoc commented:
@erik-meuwese-topicus Thanks for catching the missing license headers. I've fixed them. Is there anything else blocking approval and merge?

erik-meuwese-topicus commented on src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:
```diff
@@ -336,6 +353,41 @@ private UnaryOperator<File> createRestorationToDiskConsumer(final MavenProject p
+        if (!Files.exists(restorationPath)) {
+            Files.createDirectories(restorationPath);
+        }
+        CacheUtils.unzip(cachedZip.toPath(), restorationPath);
```
`CacheUtils.unzip(cachedZip.toPath(), restorationPath)` misses the last parameter `preservePermissions` whether to preserve Unix file permissions in the zip.

elharo commented on src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:
```diff
@@ -138,33 +139,58 @@ public void execute(
+                        cacheController.stagePreExistingArtifacts(session, project);
+                    } catch (IOException e) {
+                        LOGGER.warn("Failed to stage pre-existing artifacts: {}", e.getMessage());
+                        // Continue build - if staging fails, we'll just cache what exists
```
Then you don't need to warn. caching should be invisible to the user. move this to debug

cowwoc replied:
Done. Try now.

cowwoc commented:
@elharo As far as I can tell, nothing in the PR's implementation is git-specific but I've committed an update that removes any git-specific terminology. Please let me know if I've missed anything. Thanks!

cowwoc commented:
This is a never-ending review cycle. What will it take to get this PR merged ...?

cowwoc commented:
@olamy Can you please investigate the latest build failure? It seems to be unrelated to this PR. I see:

> java.io.FileNotFoundException: /Users/runner/.m2/settings-security.xml (No such file or directory)

and

> org.apache.http.client.HttpResponseException: status code: 500, reason phrase: Internal Server Error (500)

cowwoc commented:
Thank you for the review @Bukama! You're right - I've added documentation for the new `maven.build.cache.cacheCompile` parameter:

- **parameters.md**: Added to the command line flags table with description and usage scenario
- **how-to.md**: Added a new section "I want to disable caching of compile-only builds" explaining when users might want to disable this feature

See commit e3d8e58.

cowwoc commented:
@Bukama Good catch. Fixed by https://github.com/apache/maven-build-cache-extension/pull/394/changes/6a65421e2221505bc633fcb067267e73f8c346b4

guylabs commented:
@cowwoc Running a build with these changes and `-Dmaven.build.cache.enabled=false` set fails the build and prints this exception:

```
[ERROR] Cache is not initialized. Actual state: DISABLED
java.lang.IllegalStateException: Cache is not initialized. Actual state: DISABLED
    at org.apache.maven.buildcache.xml.CacheConfigImpl.checkInitializedState(CacheConfigImpl.java:648)
    at org.apache.maven.buildcache.xml.CacheConfigImpl.getAttachedOutputs(CacheConfigImpl.java:582)
    at org.apache.maven.buildcache.CacheControllerImpl.collectCachedArtifactPaths(CacheControllerImpl.java:1315)
    at org.apache.maven.buildcache.CacheControllerImpl.stagePreExistingArtifacts(CacheControllerImpl.java:1249)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute(BuildCacheMojosExecutionStrategy.java:161)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:110)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:61)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleStarter.execute(DefaultLifecycleStarter.java:123)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:310)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:452)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:134)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:64)
    at org.codehaus.groovy.vmplugin.v8.IndyInterface.selectMethod(IndyInterface.java:355)
    at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
    at com.gradle.testlib.use.runner.Embedded3xLauncher.doMain(Embedded3xLauncher.groovy:315)
    at com.gradle.testlib.use.runner.Embedded3xLauncher$_run_closure3.doCall(Embedded3xLauncher.groovy:282)
    at com.gradle.testlib.use.runner.Embedded3xLauncher$_run_closure3.call(Embedded3xLauncher.groovy)
    at java.util.concurrent.FutureTask.run(FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1583)
[ERROR] 
```

Running it together with `-Dmaven.build.cache.cacheCompile=false` will not fail the build.

Would it be possible to include this use case as well, as this is an official configuration option? Thanks!

cowwoc commented:
@guylabs Good catch! I've created https://github.com/apache/maven-build-cache-extension/pull/424 to fix this bug. Thanks for the report.
